proposals/gc/MVP.md

```diff
@@ -132,33 +172,35 @@ In addition to the [existing rules](https://github.com/WebAssembly/function-refe
+* Validation requires that each RTT's parent type is a representative of a static supertype; runtime subtyping hence is a sub-relation of static subtyping (a graph with fewer nodes and edges).
 
 * At the same time, runtime subtyping forms a linear hierarchy such that the relation can be checked efficiently using standard implementation techniques (the runtime subtype hierarchy is a tree-shaped graph).
 
-Note: RTT values correspond to type descriptors or "shape" objects as they exist in various engines. Moreover, runtime casts along the hierachy encoded in these values can be implemented in an engine efficiently by including a vector of its (direct and indirect) super-RTTs in each RTT value (with itself as the last entry). The value `<n>` then denotes the length of this vector. A subtype check between two RTT values can be implemented as follows using such a representation. Assume RTT value v1 has static type `(rtt n1 t1)` and v2 has type `(rtt n2 t2)`. To check whether v1 denotes a sub-RTT of v2, first verify that `n1 >= n2`. Then compare v2 to the n2-th entry in v1's supertype vector. If they are equal, v1 is a sub-RTT. For casts, the static type of v1 (taken from the object to cast) is not known at compile time, so `n1 >= n2` becomes a dynamic check as well.
+Note: RTT values correspond to type descriptors or "shape" objects as they exist in various engines. Moreover, runtime casts along the hierachy encoded in these values can be implemented in an engine efficiently by using well-known techniques such as including a vector of its (direct and indirect) super-RTTs in each RTT value (with itself as the last entry). The value `<n>` then denotes the length of this vector. A subtype check between two RTT values can be implemented as follows using such a representation. Assume RTT value v1 has static type `(rtt n1? $t1)` and v2 has type `(rtt n2? $t2)`. To check whether v1 denotes a sub-RTT of v2, first verify that `n1 >= n2` -- if both `n1` and `n2` are known statically, this can be performed at compile time; if either is not statically known, it has to be read from the respective RTT value dynamically, and `n1 >= n2` becomes a dynamic check. Then compare v2 to the n2-th entry in v1's supertype vector. If they are equal, v1 is a sub-RTT.
+In the case of actual casts, the static type of RTT v1 (taken from the  to cast) is not known at compile time, so `n1` is dynamic as well.
+(Note that `$t2` and `$t2` are not relevant for the dynamic semantics, but merely for validation.)
 
 Example: Consider three types and corresponding RTTs:
 ```
 (type $A (struct))
 (type $B (struct (field i32)))
 (type $C (struct (field i32 i64)))
 
-(global $rttA (rtt 1 $A) (rtt.sub $A (rtt.canon any)))
-(global $rttB (rtt 2 $B) (rtt.sub $B (global.get $rttA)))
-(global $rttC (rtt 3 $C) (rtt.sub $C (global.get $rttB)))
+(global $rttA (rtt 0 $A) (rtt.canon $A))
```

Comment:
Is it possible to create a depth-0 RTT without using `rtt.canon`? Otherwise, e.g. in a class-based language, two disjoint roots of different class hierarchies that just happen to be empty will get lumped together by using `rtt.canon`. They'd have to declare a superfluous empty struct to use as their parent to get different hierarchies started.

Reply:
Ok, well I misunderstood then. I was under the impression that `rtt.sub` was generative--i.e. it would generate a new RTT value each time. If RTT's are also structural, they are effectively useless for encoding a nominal class hierarchy. That's pretty much a deal breaker unless I am missing something else.

Reply:
Thus the TODO. Though let me point out the obvious: even with generative RTTs added, Wasm-level RTTs and casts will never be sufficient to directly implement the majority of source-language nominal hierarchies and casts -- for the same reason that Wasm's static type can never encode all source type systems. You can piggy-back source casts on Wasm casts in some limited cases, but don't expect that to be possible in every case. In the general case, you have to implement your own tagging system in user space. (And that makes me wonder whether generative RTTs are even worth adding to the MVP.)

Reply:
I'd argue that generative RTT values are actually cheaper and simpler to implement in an engine than structural RTT values, and I can verify that they are enough to implement the ubiquitous class casts in Java such that a wasm-level cast implies a Java-level cast, even, and perhaps especially, with classloaders. (They are not sufficient for interface or array casts, however). Without generative RTTs then a Java implementation is forced into doing both a wasm-level cast and a Java-level check, because with structural RTTs there is no way to make a wasm-level cast imply a Java-level check.

Reply:
Agreed, but let's keep generative RTTs for a different PR. This one does not change that aspect, so should be orthogonal.

Reply:
Ok.